### PR TITLE
Admin Router: Prevent reusing tcp sockets by AR's cache code [master]

### PR DESCRIPTION
**Pull request by vespian**
## High-level description

This PR prevents AR's cache code from re-using TCP connections while refreshing cache. This may result in AR trying to fetch e.g. `/mesos/state-summary` from a master that already abdicated which in turn results in failure to refresh the cache.

Please check the linked Jira for detailed analysis and discussion of the bug.

On top of this chance few small enchancements and bugfixes were done to test harness in order to be able to write a test for this change:
* mesos mock should always respond using HTTP/1.1
* the `Content-Length` field should be properly set while sending the response to client
* add `pytest-repeat` package to simplify re-running tests multiple times

## Corresponding DC/OS tickets (obligatory)

  - https://jira.mesosphere.com/browse/SOAK-68 - `soak111: Marathon UI not loading, 404 returned by adminrouter`

## Related PRs:

DC/OS EE PR: https://github.com/mesosphere/dcos-enterprise/pull/2331

### CODE REVIEW

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot label ready for review

**Comment by vespian** [comment]
@jgehrcke @gpaul  Ready for review modulo CI.

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/arm https://jira.mesosphere.com/browse/DCOS_OSS-2115

**Comment by vespian** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/arm https://jira.mesosphere.com/browse/DCOS-21476

**Comment by sambatyon** [comment]
Could you write which tests are failing and the jiras tracking them? I know `test_networking.test_vip` is flaky but I have never seen `test_rlimit.test_if_rlimits_can_be_used` or `test_sysctl.test_if_default_systctls_are_set` failing.

**Comment by vespian** [comment]
> Could you write which tests are failing and the jiras tracking them? I know test_networking.test_vip is flaky but I have never seen test_rlimit.test_if_rlimits_can_be_used or test_sysctl.test_if_default_systctls_are_set failing.

@sambatyon we do not usually write the names of individual tests, e.g.
* https://github.com/dcos/dcos/pull/2493
* https://github.com/dcos/dcos/pull/2426
* https://github.com/dcos/dcos/pull/2496
as the issue owner will most probably open the PRs by clicking the link from the flakiness Jira issue instead of going through the individual PRs. I doubt anybody will be scanning individual PRs one month from now on.  I also find this convenient to just do the override instead of listing all the test names, corresponding CIs that failed, and jiras that track them - this allows me to limit the amount of typing and move faster to the next task.

But in order to shorten this conversation:
* `test_rlimit.test_if_rlimits_can_be_used` failed in CI task `
teamcity/dcos/test/azure/arm` and is tracked in https://jira.mesosphere.com/browse/DCOS-21476
* ` test_sysctl.test_if_default_systctls_are_set` failed in CI task `
teamcity/dcos/test/azure/arm` and is tracked in https://jira.mesosphere.com/browse/DCOS-21476
* ` test_networking.test_vip[Container_NONE-Network_USER-Network_USER]` failed in CI task 
teamcity/dcos/test/docker  and is tracked in https://jira.mesosphere.com/browse/DCOS_OSS-2115

**Comment by gpaul** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/docker https://jira.mesosphere.com/browse/DCOS_OSS-2115

File: packages/adminrouter/extra/src/lib/cache.lua [hunk 1]
```diff
@@ -59,7 +59,15 @@ end
 
 
 local function request(url, accept_404_reply, auth_token)
-    local headers = {["User-Agent"] = "Master Admin Router"}
+    -- We need to make sure that Nginx does not reuse the TCP connection here,
```
Comment (gpaul) [inline packages/adminrouter/extra/src/lib/cache.lua]: Very interesting bug!

Comment (sambatyon) [reply packages/adminrouter/extra/src/lib/cache.lua]: I was wondering about it too, since most of our code has been moved towards not closing connections.

**Comment by vespian** [comment]
A followup for https://github.com/dcos/dcos/pull/2511#issuecomment-369516609

@sambatyon's main concern here was that it seemed that I am trying to override my own PR which will not result in properly linking the failure to a Jira. My intention here was to make it easier for other owners by providing them easily copy-pastable text so that the interruption for them is minimal.

Thanks @sambatyon  for reaching out and clarifying it!

**Comment by gpaul** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/arm https://jira.mesosphere.com/browse/DCOS-21476

**Pull request merged**